qutebrowser/browser/greasemonkey.py

```diff
@@ -171,23 +194,110 @@ def load_scripts(self):
                 if not script.name:
                     script.name = script_filename
 
-                if script.run_at == 'document-start':
-                    self._run_start.append(script)
-                elif script.run_at == 'document-end':
-                    self._run_end.append(script)
-                elif script.run_at == 'document-idle':
-                    self._run_idle.append(script)
+                if script.requires:
+                    log.greasemonkey.debug(("Deferring script until "
+                                            "requirements are "
+                                            "fullfilled: {}")
+                                           .format(script.name))
+                    self._get_required_scripts(script, force)
                 else:
-                    log.greasemonkey.warning("Script {} has invalid run-at "
-                                             "defined, defaulting to "
-                                             "document-end"
-                                             .format(script_path))
-                    # Default as per
-                    # https://wiki.greasespot.net/Metadata_Block#.40run-at
-                    self._run_end.append(script)
-                log.greasemonkey.debug("Loaded script: {}".format(script.name))
+                    self._add_script(script)
+
         self.scripts_reloaded.emit()
 
+    def _add_script(self, script):
+        if script.run_at == 'document-start':
+            self._run_start.append(script)
+        elif script.run_at == 'document-end':
+            self._run_end.append(script)
+        elif script.run_at == 'document-idle':
+            self._run_idle.append(script)
+        else:
+            log.greasemonkey.warning("Script {} has invalid run-at "
+                                     "defined, defaulting to "
+                                     "document-end"
+                                     .format(script.name))
+            # Default as per
+            # https://wiki.greasespot.net/Metadata_Block#.40run-at
+            self._run_end.append(script)
+        log.greasemonkey.debug("Loaded script: {}".format(script.name))
+
+    def _required_url_to_file_path(self, url):
+        # Save to a more readable name
+        # cf https://stackoverflow.com/questions/295135/turn-a-string-into-a-valid-filename
+        # replace everything except hyphen, word characters, underscore and dot
+        # with and underscore.
+        name = re.sub(r'(?u)[^-\w.]', '_', url)
+        requires_dir = os.path.join(_scripts_dir(), 'requires')
+        if not os.path.exists(requires_dir):
+            os.mkdir(requires_dir)
+        return os.path.join(requires_dir, name)
+
+    def _on_required_download_finished(self, script, download=None,
+                                       quiet=False):
```

Comment:
The `download=None, quiet=False` thing makes me wonder whether it'd be cleaner to split this up into two functions, one handling a download and one assembling stuff?